Userland/Libraries/LibJS/SourceCode.cpp

```diff
@@ -32,18 +33,50 @@ String const& SourceCode::code() const
     return m_code;
 }
 
+void SourceCode::compute_line_break_offsets() const
+{
+    if (m_code.is_empty())
+        return;
+
+    bool previous_code_point_was_carriage_return = false;
+    Utf8View view(m_code.view());
+    for (auto it = view.begin(); it != view.end(); ++it) {
+        u32 code_point = *it;
+        bool is_line_terminator = code_point == '\r' || (code_point == '\n' && !previous_code_point_was_carriage_return) || code_point == LINE_SEPARATOR || code_point == PARAGRAPH_SEPARATOR;
+        previous_code_point_was_carriage_return = code_point == '\r';
+
+        if (is_line_terminator)
+            m_line_break_offsets.append(view.byte_offset_of(it));
+    }
+}
+
 SourceRange SourceCode::range_from_offsets(u32 start_offset, u32 end_offset) const
 {
+    if (m_code.is_empty())
+        return { *this, {}, {} };
+
+    if (m_line_break_offsets.is_empty())
```

Comment:
Not sure this is sufficient, especially on the web where a lot of JS is minified into a single line. Maybe make it an `Optional<Vector<size_t>>`?

Reply:
Great point, will fix!